pkg/netpol/eval/internal/k8s/netpol.go

```diff
@@ -200,27 +202,42 @@ func (np *NetworkPolicy) ruleSelectsPeer(rulePeers []netv1.NetworkPolicyPeer, pe
 			// peer is a pod
 			peerMatchesPodSelector := false
 			peerMatchesNamespaceSelector := false
+			var err error
+			var selector labels.Selector
 			if rulePeers[i].NamespaceSelector == nil {
 				peerMatchesNamespaceSelector = (np.ObjectMeta.Namespace == peer.GetPeerPod().Namespace)
 			} else {
-				selector, err := np.parseNetpolLabelSelector(rulePeers[i].NamespaceSelector)
+				peerNamespace := peer.GetPeerNamespace()
+				// checking if a selector matches labels by peer type; since representative peers selectors may need special handling
+				if hasRepresentativePod(peer) {
+					// representative peer is inferred from a rule with its/same labelSelector.
+					// note that if the namespaceSelector in the rule is nil, we don't get here,
+					// since that means the peer is in same namespace of the policy, and this was checked above
+					peerMatchesNamespaceSelector, err = SelectorsFullMatch(rulePeers[i].NamespaceSelector, peer.GetPeerPod().RepresentativeNsLabelSelector)
+				} else {
+					selector, err = np.parseNetpolLabelSelector(rulePeers[i].NamespaceSelector)
+					peerMatchesNamespaceSelector = selector.Matches(labels.Set(peerNamespace.Labels))
+				}
 				if err != nil {
 					return false, err
 				}
-				peerNamespace := peer.GetPeerNamespace()
-				peerMatchesNamespaceSelector = selector.Matches(labels.Set(peerNamespace.Labels))
 			}
 			if !peerMatchesNamespaceSelector {
 				continue // skip to next peerObj
 			}
 			if rulePeers[i].PodSelector == nil {
 				peerMatchesPodSelector = true
 			} else {
-				selector, err := np.parseNetpolLabelSelector(rulePeers[i].PodSelector)
+				// checking if a selector matches labels by peer type; since representative peers selectors may need special handling
```

Comment:
can capture this common code in a separate function so it is not duplicated?

Reply:
done